tf_encrypted/keras/layers/pooling.py

```diff
@@ -167,3 +167,104 @@ def __init__(self,
+
+  def compute_output_shape(self, input_shape):
+    if self.data_format == 'channels_last':
+      output_shape = [input_shape[0], input_shape[3]]
+    else:
+      output_shape = [input_shape[0], input_shape[1]]
+
+    return output_shape
+
+  def call(self, inputs):
+    raise NotImplementedError
+
+class GlobalAveragePooling2D(GlobalPooling2D):
+  """Global average pooling operation for spatial data.
+
+  Arguments:
+      data_format: A string,
+          one of `channels_last` (default) or `channels_first`.
+          The ordering of the dimensions in the inputs.
+          `channels_last` corresponds to inputs with shape
+          `(batch, height, width, channels)` while `channels_first`
+          corresponds to inputs with shape
+          `(batch, channels, height, width)`.
+          It defaults to the `image_data_format` value found in your
+          Keras config file at `~/.keras/keras.json`.
+          If you never set it, then it will be "channels_last".
+
+  Input shape:
+      - If `data_format='channels_last'`:
+          4D tensor with shape:
+          `(batch_size, rows, cols, channels)`
+      - If `data_format='channels_first'`:
+          4D tensor with shape:
+          `(batch_size, channels, rows, cols)`
+
+  Output shape:
+      2D tensor with shape:
+      `(batch_size, channels)`
+  """
+  def build(self, input_shape):
+    if self.data_format == 'channels_last':
+      _, h_in, w_in, _ = input_shape
+    else:
+      _, _, h_in, w_in = input_shape
+
+    self.scalar = 1 / int(h_in * w_in)
+
+  def call(self, inputs):
+    if self.data_format == 'channels_last':
+      x_reduced = inputs.reduce_sum(axis=2).reduce_sum(axis=1)
+    else:
+      x_reduced = inputs.reduce_sum(axis=3).reduce_sum(axis=2)
+
+    return x_reduced * self.scalar
+
+
+class GlobalMaxPooling2D(GlobalPooling2D):
+  """Global max pooling operation for spatial data.
+
+  Arguments:
+      data_format: A string,
+          one of `channels_last` (default) or `channels_first`.
+          The ordering of the dimensions in the inputs.
+          `channels_last` corresponds to inputs with shape
+          `(batch, height, width, channels)` while `channels_first`
+          corresponds to inputs with shape
+          `(batch, channels, height, width)`.
+          It defaults to the `image_data_format` value found in your
+          Keras config file at `~/.keras/keras.json`.
+          If you never set it, then it will be "channels_last".
+
+  Input shape:
+      - If `data_format='channels_last'`:
+          4D tensor with shape:
+          `(batch_size, rows, cols, channels)`
+      - If `data_format='channels_first'`:
+          4D tensor with shape:
+          `(batch_size, channels, rows, cols)`
+
+  Output shape:
+      2D tensor with shape:
+      `(batch_size, channels)`
+  """
+
+  def call(self, inputs):
+    if self.data_format == 'channels_last':
+      x_reduced = self.prot.reduce_max(inputs, axis=2)
```

Comment:
is `reduce_max` not implemented as a tensor method? if so, we should open a separate issue to track that

Reply:
it isn't implemented in pond, and when I called it using the tensor (inputs.reduce_max) i get an error that says that pond doesn't have reduce max

Reply:
I think this should be covered by our refactoring plans, so nvm on tracking with an issue